### PR TITLE
nlopt 2.7.1

### PR DESCRIPTION
**Pull request by skupr-anaconda**
**Destination channel:** {defaults}

### Links

- [PKG-4482](https://anaconda.atlassian.net/browse/PKG-4482) 
- [Upstream repository](https://github.com/stevengj/nlopt/tree/v2.7.1)
- [Upstream changelog/diff](https://github.com/stevengj/nlopt/compare/nlopt-2.4.2...v2.7.1)

### Explanation of changes:

- `nlopt` is a required dependency of `r-nloptr` (**R** interface) so the goal is to build a **non**-python interface. 
- Refresh the recipe based on the conda-forge's recipe
- Reset the build number to `0`
- Add `"$RPATH/ld64.so.1"` to `missing_dso_whitelist` on `s390x` 
- Add `m2-patch`/`patch` to `build`
- Disable the **Python** interface on `win64`: 
  - add the flag `-DNLOPT_PYTHON=OFF` to `bld.bat` and `build.sh`. See **Notes** and **comments** below.
- Add test commands
- Update `home` url
- Add dev & doc urls
- Update `license` name
- Add `license_family`
- Add `description`

### Notes:

- I wonder if we need to build `nlopt` with **python** on `win64` (???). Docs says:
`There are also interfaces for C++, Fortran, Python, Matlab or GNU Octave, OCaml, GNU Guile, GNU R, Lua, Rust, and Julia.`. But we do not need Python for win64 as we will build an R interface.

- `r-nloptr` requires `nlopt`. Around 350 R packages rely on `r-nloptr`.
- I've tested `r-nloptr 2.0.3` and  the downstream package `r-btydplus 1.2.0`  with this `nlopt 2.7.1` locally on win64 (using `MSYS2`/`UCRT64` toolchains). There is still `python` in r-nloptr's host and test environments I guess it;'s pulled by other dependency

[PKG-4482]: https://anaconda.atlassian.net/browse/PKG-4482?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by skupr-anaconda** [comment]
An error on `osx`:
`   ['  ERROR (nlopt,lib/python3.8/site-packages/_nlopt.cpython-38-darwin.so): /private/var/folders/k1/30mswbxs7r1g6zwn8y4fyt500000gp/T/abs_d13lhnrnam/croot/nlopt_1713347978677/work/build/libnlopt.0.dylib not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']`

File: recipe/bld.bat [hunk 1]
```diff
@@ -3,9 +3,11 @@ mkdir build && cd build
 
 set CMAKE_CONFIG="Release"
 
+REM nlopt is used by r-nloptr so a Python interface isn't needed on win64
 cmake -LAH -G"NMake Makefiles"                     ^
   -DCMAKE_PREFIX_PATH="%LIBRARY_PREFIX%"           ^
   -DCMAKE_INSTALL_PREFIX="%LIBRARY_PREFIX%"        ^
+  -DNLOPT_PYTHON=OFF ^
```
Comment (ifitchet) [inline recipe/bld.bat]: @skupr-anaconda can you explain why we don't want the Python interfaces for `win-64`?

File: recipe/build.sh [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
-  --without-guile
-
-make -j${CPU_COUNT} ${VERBOSE_AT}
-make install ${VERBOSE_AT}
+if [ "$(uname)" = "Darwin" ]; then
+  # osx uses a temp directory which messes up with cmake finding libraries.
+  cd "${SRC_DIR}" || exit
+fi
+
+mkdir build && cd build || exit
```
Comment (ifitchet) [inline recipe/build.sh]: Bash scripts should have `set -euo pipeline` at the start to handle commands that fail automatically.
(For existing scripts, like `r-base`'s `build.sh`, the `-u` flag may not be welcome.)

```suggestion

set -euo pipeline

if [ "$(uname)" = "Darwin" ]; then
  # osx uses a temp directory which messes up with cmake finding libraries.
  cd "${SRC_DIR}"
fi

mkdir build
cd build

```

File: recipe/build.sh [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
+cmake \
+  -DCMAKE_PREFIX_PATH=${PREFIX} \
+  -DCMAKE_INSTALL_PREFIX=${PREFIX} \
+  -DCMAKE_INSTALL_LIBDIR=lib \
+  -DNLOPT_GUILE=OFF -DNLOPT_MATLAB=OFF -DNLOPT_OCTAVE=OFF \
```
Comment (ifitchet) [inline recipe/build.sh]: Why do we disable guile, matlab and octave here but not for Windows?

**Review by ifitchet — CHANGES_REQUESTED** [review]

**Comment by JeanChristopheMorinPerso** [comment]
>Explanation of changes:
>* Refresh the recipe based on the conda-forge's recipe
>* Reset the build number to 0
>* Add "$RPATH/ld64.so.1" to missing_dso_whitelist on s390x
>* Add m2-patch/patch to build
>* Disable the Python interface on win64:
>    * add the flag -DNLOPT_PYTHON=OFF to bld.bat
>    * add the selector # [not win] to host and run sections
>* Add test commands
>* Update home url
>* Add dev & doc urls
>* Update license name
>* Add license_family
>* Add description

@skupr-anaconda Could it be possible in the future to use the "explanation" section to explain the changes more in detail? Your first item is a good explanation of what you did (Refresh the recipe based on the conda-forge's recipe). The rest is just describing the diff as is and makes it harder to see what are the changes in general (things that are harder to infer from the diff).

Also, is the Jira ID correct? It links to an opencv ticket while the notes seem to imply that this PR is to allow building `r-nloptr`.

>I wonder if we need to build nlopt with python on win64 (???). Docs says:
There are also interfaces for C++, Fortran, Python, Matlab or GNU Octave, OCaml, GNU Guile, GNU R, Lua, Rust, and Julia.. But we do not need Python for win64 as we will build an R interface.

What is the goal of the PR? If the goal is to build `nloptr`, then I would say that we don't need the python bindings. And taking a look at our dependencies, `r-nloptr` is the only package that depends on `nlopt`. I would also mirror that decision on other platforms too.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,49 +1,70 @@
...
+    - if not exist %PREFIX%/Library/lib/nlopt.lib exit 1      # [win]
+    - if not exist %PREFIX%/Library/bin/nlopt.dll exit 1      # [win]
 
 about:
-  home: http://ab-initio.mit.edu/nlopt
-  license: LGPL 2.1
+  home: https://nlopt.readthedocs.io
+  doc_url: https://nlopt.readthedocs.io
+  dev_url: https://github.com/stevengj/nlopt/
+  license: LGPL-2.1-or-later AND MIT
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]:
```suggestion
  license: LGPL-2.1-or-later AND MIT AND BSD-3-Clause
```
https://github.com/stevengj/nlopt/tree/v2.7.1/src/algs/slsqp is BSD-3-clause

File: recipe/build.sh (hunk 3 above)
Comment (JeanChristopheMorinPerso) [reply recipe/build.sh]: I would probably disable python. See my comment at https://github.com/AnacondaRecipes/nlopt-feedstock/pull/1#issuecomment-2062007729. And yeah, if we disable something on unix, I think we should also disable it on Windows. Explicit is better than implicit.

File: recipe/bld.bat (hunk 1 above)
Comment (skupr-anaconda) [reply recipe/bld.bat]: Every supported programming language can have its interface of `nlopt` but creating the **Python** interface will require `python` in `host` and at `runtime`. That means then we need to build `nlopt` for all `python` versions (**four** at this moment). Hence, `r-nloptr` will be built for all 4 Python versions and the same applies to all 350 downstream R packages. It's a lot of redundancy. And R packages seem not to require the Python interface.

File: recipe/build.sh (hunk 2 above)
Comment (skupr-anaconda) [reply recipe/build.sh]: Fixed

File: recipe/build.sh (hunk 3 above)
Comment (skupr-anaconda) [reply recipe/build.sh]: That's fair. Fixed it.

File: recipe/meta.yaml (hunk 4 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml]: Thanks! I've extended `license_file` by adding all explicit licenses in the `src/algs` folder:
```
  license_file:
    - COPYING
    - src/algs/cobyla/COPYRIGHT
    - src/algs/direct/COPYING
    - src/algs/esch/COPYRIGHT
    - src/algs/luksan/COPYRIGHT
    - src/algs/newuoa/COPYRIGHT
    - src/algs/slsqp/COPYRIGHT
    - src/algs/stogo/COPYRIGHT
```

File: recipe/build.sh (hunk 2 above)
Comment (skupr-anaconda) [reply recipe/build.sh]: `set -euo pipefail`

File: recipe/build.sh [hunk 5]
```diff
@@ -1,15 +1,30 @@
...
-make -j${CPU_COUNT} ${VERBOSE_AT}
-make install ${VERBOSE_AT}
+set -euo pipefail
+
+if [ "$(uname)" = "Darwin" ]; then
+  # osx uses a temp directory which messes up with cmake finding libraries.
+  cd "${SRC_DIR}" || exit
+fi
+
+mkdir build && cd build || exit
```
Comment (ifitchet) [inline recipe/build.sh]: 😆 the point about `set -e` it that `cd`, `mkdir` etc. will cause the script to exit if they fail.  You don't need `|| exit [1]`

```suggestion
  cd "${SRC_DIR}"
fi

mkdir build
cd build
```

Also, putting one command per line (`mkdir` ... `cd`) makes a script more readable for the nervous -- and you've captured the "logical and" (`&&`) with `set -e`, above.

**Review by ifitchet — CHANGES_REQUESTED** [review]

**Comment by skupr-anaconda** [comment]
@JeanChristopheMorinPerso @ifitchet Before this build for the `defaults` channel our `r-nloptr` relied on the `ucrt64-nlopt` package (see the artifact https://anaconda.org/skupr/ucrt64-nlopt/files). 
It's repacking of the MSYS2 package `mingw-w64-nlopt` (see the recipe https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-nlopt/PKGBUILD) and our `mingw-w64-nlotp`'s recipe didn't have Python in host and run but `recipe/repack.bat` copied the python files `_nlopt.pyd` and `nlopt.py` to `Library/ucrt64/lib/site-packages/`:
```
set src="%SRC_DIR%\%PKG_NAME%"
set dst="%PREFIX%\Library"
robocopy /E /NFL /NDL /NP /NJH /NJS %src% %dst%
```
But while compiling the package `r-nloptr `doesn't invoke Python. 
```
g++ -shared -s -o nloptr.dll tmp.def init_nloptr.o nloptr.o test-C-API.o test-runner.o -lnlopt -LC:/Users/dev/miniconda3/conda-bld/r-nloptr_1713429374939/_h_env/Library/lib/x64 -LC:/Users/dev/miniconda3/conda-bld/r-nloptr_1713429374939/_h_env/Library/lib -LC:/Users/dev/miniconda3/conda-bld/r-nloptr_1713429374939/_h_env/lib/R/bin/x64 -lR
installing to C:/Users/dev/miniconda3/conda-bld/r-nloptr_1713429374939/_h_env/Lib/R/library/00LOCK-work/00new/nloptr/libs/x64
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* MD5 sums
packaged installation of 'nloptr' as nloptr_2.0.3.zip
* DONE (nloptr)
```

Its docs don't say about Python too:
- https://github.com/astamm/nloptr/tree/master?tab=readme-ov-file#installation
- https://github.com/astamm/nloptr/blob/master/DESCRIPTION#L40-L41
- https://github.com/astamm/nloptr/blob/master/src/Makevars.win
- https://github.com/astamm/nloptr/blob/master/src/Makevars.in 

So I've decided to disable compiling `nlopt`'s Python stuff on `unix` and `win` platforms.

I've tested examples from https://cloud.r-project.org/web/packages/nloptr/nloptr.pdf and they work fine on `win64`.

File: recipe/meta.yaml [hunk 6]
```diff
@@ -1,49 +1,69 @@
...
   build:
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
-    - autoconf
-    - automake
-    - libtool
-    - cmake         # [win]
+    - cmake
+    - make      # [unix]
     - swig
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml:27]: Is swig needed now that you disabled all the bindings?

**Review by JeanChristopheMorinPerso — APPROVED** [review]
I left a small comment regarding swig, but apart from that, it lokos good to me.

File: recipe/meta.yaml (hunk 6 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml:27]: No, we do not need it anymore as swig is required to generate python and swig bindings

Comment (skupr-anaconda) [reply recipe/meta.yaml:27]: I disabled swig

**Review by ifitchet — APPROVED** [review]
Let's do it!

And wait for the masses to complain there are no Python bindings...

**Comment by skupr-anaconda** [comment]
> Let's do it!
> 
> And wait for the masses to complain there are no Python bindings...

There is a third-party Python package for `nlopt` https://github.com/DanielBok/nlopt-python (https://pypi.org/project/nlopt/)

**Pull request merged**